### DISCUSSION TITLE
ways to reduce BVH memory for large collection of spheres...

### DISCUSSION THREAD

**Discussion by timsandstrom** [RenderKit/ospray#572] [Ideas]
I'm trying to render many spheres, currently about ~2B.

I've set the 'compactMode' flag but am still running out of memory on a 192GB machine :)

Are there better flags or perhaps I can select a more thrifty BVH?

-Tim

**Comment by johguenther** [comment]
One trick to use is to let OSPRay share the sphere data with Embree, which means to have a layout `[x, y, z, r]`, i.e., center position _and_ per-sphere radius. Still, two OSPRay _shared_ data objects needs to be created by the app, one for `sphere.position`, one for `sphere.radius`, using a byte-stride of `sizeof(vec4f)=16`.

With that I can render 2B spheres with using about 100GB of RAM, the BVH took about 48GB (depending a bit on the distribution of the spheres, but regardless of whether `compactMode` or not).

To further reduce the data we'd need to use compression / quantization, or revive the stale PKDtree module https://github.com/ingowald/ospray-module-pkd
